### ISSUE TITLE
bind-backend regression 3.x -> 4.x with multiple backends slips in data from wrong zone

### ISSUE THREAD

**Issue by ZaphodB** [PowerDNS/pdns#4907]
- Program: Authoritative 
 - Issue type: Bug report 

### Short description

3 backends in order:
 - bind for service checks loading zone "localhost"
 - gmysql DB pdns_slave for slave zones
 - gmysql DB pdns for NATIVE zones

After upgrading from 3.4.1-4+deb8u6 to 4.0.2-1pdns.jessie PowerDNS Server returned zone "localhost" data along with the SOA for the first slave zone in pdns_slave DB although that slave zone does not have anything like localhost in it.

with
`launch=bind:servicecheck,gmysql:slave,gmysql:master`

```
➜  ~ dig any collins.kg @wv-adns2-ha2.wavecloud.de
;; Truncated, retrying in TCP mode.

; <<>> DiG 9.8.3-P1 <<>> any collins.kg @wv-adns2-ha2.wavecloud.de
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 34362
;; flags: qr aa rd; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;collins.kg.			IN	ANY

;; ANSWER SECTION:
localhost.		86400	IN	NS	localhost.
localhost.		86400	IN	A	127.0.0.1
collins.kg.		3600	IN	SOA	ns1.wavedns.de. it-guys.project-collins.com. 2017011201 600 3600 604800 600

;; Query time: 36 msec
;; SERVER: 185.22.222.15#53(185.22.222.15)
;; WHEN: Fri Jan 13 19:47:57 2017
;; MSG SIZE  rcvd: 144
```

with
`launch=gmysql:slave,gmysql:master,bind:servicecheck`

```
➜  ~ dig any collins.kg @wv-adns2-ha2.wavecloud.de
;; Truncated, retrying in TCP mode.

; <<>> DiG 9.8.3-P1 <<>> any collins.kg @wv-adns2-ha2.wavecloud.de
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 41358
;; flags: qr aa rd; QUERY: 1, ANSWER: 10, AUTHORITY: 0, ADDITIONAL: 2
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;collins.kg.			IN	ANY

;; ANSWER SECTION:
collins.kg.		3600	IN	A	193.192.41.26
collins.kg.		3600	IN	MX	10 ASPMX.L.GOOGLE.COM.
collins.kg.		3600	IN	MX	20 ALT1.ASPMX.L.GOOGLE.COM.
collins.kg.		3600	IN	MX	20 ALT2.ASPMX.L.GOOGLE.COM.
collins.kg.		3600	IN	MX	30 ASPMX2.GOOGLEMAIL.COM.
collins.kg.		3600	IN	MX	30 ASPMX3.GOOGLEMAIL.COM.
collins.kg.		3600	IN	NS	ns1.wavedns.de.
collins.kg.		3600	IN	NS	ns7.wavedns.de.
collins.kg.		300	IN	TXT	"v=spf1 include:_spf.google.com ip4:193.192.41.0/24 ~all"
collins.kg.		3600	IN	SOA	ns1.wavedns.de. it-guys.project-collins.COM. 2017011201 600 3600 604800 600

;; ADDITIONAL SECTION:
ns1.wavedns.de.		86400	IN	A	185.22.221.152
ns7.wavedns.de.		86400	IN	A	185.22.222.32

;; Query time: 40 msec
;; SERVER: 185.22.222.15#53(185.22.222.15)
;; WHEN: Fri Jan 13 19:48:38 2017
;; MSG SIZE  rcvd: 383
```

### Environment

 - Operating system: Debian stable (8.6)
 - Software version: 3.4.1-4+deb8u6 upgraded to 4.0.2-1pdns.jessie
 - Software source: PowerDNS repository

### Steps to reproduce

pdns.conf:
```
allow-axfr-ips=127.0.0.1,::1
allow-recursion=127.0.0.1
cache-ttl=20
config-dir=/etc/powerdns
daemon=yes
default-soa-name=ns1.wavedns.de
disable-axfr=no
disable-tcp=no
distributor-threads=3
guardian=yes
launch=gmysql:slave,gmysql:master,bind:servicecheck
bind-servicecheck-config=/etc/powerdns/bind.conf
gmysql-master-host=127.0.0.1
gmysql-master-port=3306
gmysql-master-dbname=pdns
gmysql-master-user=pdns
gmysql-master-password=password
gmysql-master-dnssec=yes
gmysql-slave-host=127.0.0.1
gmysql-slave-port=3306
gmysql-slave-dbname=pdns_slave
gmysql-slave-user=pdns
gmysql-slave-password=password
gmysql-slave-dnssec=no
local-address=185.22.222.15, 185.22.222.32, 127.0.0.1
local-port=53
master=no
negquery-cache-ttl=60
query-cache-ttl=20
query-local-address=185.22.222.15
receiver-threads=1
recursive-cache-ttl=10
setgid=pdns
setuid=pdns
slave=no
slave-cycle-interval=60
socket-dir=/var/run
webserver=yes
webserver-address=127.0.0.1
webserver-port=8081
version-string=powerdns
```

bind.conf:
`zone "localhost" in { type master; file "/etc/powerdns/localhost.zone"; };`

localhost.zone:
```
$TTL    1D
$ORIGIN localhost.
@       1D      IN   SOA     localhost. root.localhost. (
                1      ; serial
                3H              ; refresh
                1H              ; retry
                1W              ; expiry
                1D )            ; minimum
        IN NS localhost.
        IN A 127.0.0.1
```

pdns_slave:
```
mysql> select * from domains;
+----+---------------------+------------------------------+------------+-------+-----------------+---------+
| id | name                | master                       | last_check | type  | notified_serial | account |
+----+---------------------+------------------------------+------------+-------+-----------------+---------+
|  1 | collins.kg          | 4.3.2.1, 1.2.3.4 | 1484332798 | SLAVE |            NULL | NULL    |
|  2 | schlampenexperte.de | 8.7.6.5               | 1484329663 | SLAVE |            NULL | NULL    |
+----+---------------------+------------------------------+------------+-------+-----------------+---------+
2 rows in set (0.00 sec)
```

```
mysql> select * from records where domain_id=1 limit 10;
+---------+-----------+------------+------+-----------------------------------------------------------------------------+------+------+-------------+-----------+------+----------+
| id      | domain_id | name       | type | content                                                                     | ttl  | prio | change_date | ordername | auth | disabled |
+---------+-----------+------------+------+-----------------------------------------------------------------------------+------+------+-------------+-----------+------+----------+
| 6683302 |         1 | collins.kg | SOA  | ns1.wavedns.de. it-guys.project-collins.com. 2017011201 600 3600 604800 600 | 3600 |    0 |        NULL | NULL      |    1 |        0 |
| 6683303 |         1 | collins.kg | A    | 193.192.41.26                                                               | 3600 |    0 |        NULL | NULL      |    1 |        0 |
| 6683304 |         1 | collins.kg | MX   | ASPMX.L.GOOGLE.COM                                                          | 3600 |   10 |        NULL | NULL      |    1 |        0 |
| 6683305 |         1 | collins.kg | MX   | ALT1.ASPMX.L.GOOGLE.COM                                                     | 3600 |   20 |        NULL | NULL      |    1 |        0 |
| 6683306 |         1 | collins.kg | MX   | ALT2.ASPMX.L.GOOGLE.COM                                                     | 3600 |   20 |        NULL | NULL      |    1 |        0 |
| 6683307 |         1 | collins.kg | MX   | ASPMX2.GOOGLEMAIL.COM                                                       | 3600 |   30 |        NULL | NULL      |    1 |        0 |
| 6683308 |         1 | collins.kg | MX   | ASPMX3.GOOGLEMAIL.COM                                                       | 3600 |   30 |        NULL | NULL      |    1 |        0 |
| 6683309 |         1 | collins.kg | NS   | ns1.wavedns.de                                                              | 3600 |    0 |        NULL | NULL      |    1 |        0 |
| 6683310 |         1 | collins.kg | NS   | ns7.wavedns.de                                                              | 3600 |    0 |        NULL | NULL      |    1 |        0 |
| 6683311 |         1 | collins.kg | TXT  | "v=spf1 include:_spf.google.com ip4:193.192.41.0/24 ~all"                   |  300 |    0 |        NULL | NULL      |    1 |        0 |
+---------+-----------+------------+------+-----------------------------------------------------------------------------+------+------+-------------+-----------+------+----------+
10 rows in set (0.00 sec)
```

### Expected behaviour

Querying zone collins.kg should not return 'localhost' data.

### Actual behaviour

Querying zone collins.kg returned 'localhost' data from bind backend.

### Other information

just
```
-launch=bind:servicecheck,gmysql:slave,gmysql:master
+launch=gmysql:slave,gmysql:master,bind:servicecheck
```
and restaring pdns does the trick. At least for collins.kg.

**Comment by ZaphodB** [comment]
Actually changing the backend launch order does not really seem to help just for ANY records at zone apex.

Sadly i still get this:
```
➜  ~ dig any sdfasdf.www.collins.kg @wv-adns2-ha2.wavecloud.de
;; Truncated, retrying in TCP mode.

; <<>> DiG 9.8.3-P1 <<>> any sdfasdf.www.collins.kg @wv-adns2-ha2.wavecloud.de
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 59972
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;sdfasdf.www.collins.kg.		IN	ANY

;; AUTHORITY SECTION:
localhost.		86400	IN	NS	localhost.

;; ADDITIONAL SECTION:
localhost.		86400	IN	A	127.0.0.1

;; Query time: 37 msec
;; SERVER: 185.22.222.15#53(185.22.222.15)
;; WHEN: Fri Jan 13 20:15:01 2017
;; MSG SIZE  rcvd: 79
```

**Comment by Habbie** [comment]
@rgacogne has identified b854d9fec6f7e5636ab4742d716c7d848e0ce0d6 as the culprit, revert is coming up I think.

**Comment by tbaschak** [comment]
I've confirmed in IRC that this is also what was affecting me, I went from 4.0.1-1pdns.jessie to 4.0.2-1pdns.jessie. gmysql and bind backend.

**Comment by zeha** [comment]
Had a quick look at the call chain there and it's likely one of these 'UeberBackend isnt a DNSBackend and you cant treat it like one' things :-(

**Comment by zeha** [comment]
BTW, this should also go away if you do not have overlapping domain_ids; bindbackend assigns those starting at 1.

**Comment by ZaphodB** [comment]
Solved with #4905 / 4.0.3 .

**Comment by Habbie** [comment]
Dis-solved in 4.0.4, sorry!

On 24 January 2017 19:49:05 CET, Stefan Schmidt <notifications@github.com> wrote:
>Solved with #4905 / 4.0.3 .
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/PowerDNS/pdns/issues/4907#issuecomment-274898100

-- 
Sent from my Android phone with K-9 Mail. Please excuse my brevity.

**Comment by ZaphodB** [comment]
@Habbie can you elaborate on this? It sounds like you are going to break this again with 4.0.4. Combined with what @zeha wrote i get the feeling that i am not supposed to have multiple backends sporting the same domain_id internally, however with bindbackend i don't have any influence on where it starts to count and https://doc.powerdns.com/md/appendix/backend-writers-guide/ only states "Each zone has an id which should be unique within the backend."
While i only run bindbackend for Loadbalancer service checks i think it likely there are other people out there to whom it matters to have bindbackend queried first.
